### PR TITLE
perf(lsp): reduce polling handles for workspace/didChangeWatchedFiles

### LINKED ISSUE #23291: LSP: high cpu usage with huge number of files with `workspace/didChangeWatchedFiles` enabled
### Problem

Noticed this because I have some huge directories in one of my repos for testing, which probably isn't very common. Excluding files not in the patterns sent by the LSP server could possibly help. Rust-analyzer sends `root_dir/**/*.rs`, `root_dir/**/Cargo.toml`, `root_dir/**/Cargo.lock`.

### Steps to reproduce using "nvim -u minimal_init.lua"

steps:
```sh
cargo new repro --bin && cd repro
mkdir test && printf '%s\n' test/{0..60000} | xargs touch
nvim -u init.lua src/main.rs
```

init.lua:
```lua
local pattern = 'rust'
local cmd = {'rust-analyzer'}
local root_markers = {'Cargo.toml'}
local settings = vim.empty_dict()
local capabilities = vim.lsp.protocol.make_client_capabilities()
-- uncommenting this solves the stutter
--capabilities.workspace.didChangeWatchedFiles.dynamicRegistration = false
vim.api.nvim_create_autocmd('FileType', {
	pattern = pattern,
	callback = function(args)
		local match = vim.fs.find(root_markers, { path = args.file, upward = true })[1]
		local root_dir = match and vim.fn.fnamemodify(match, ':p:h') or nil
		vim.lsp.start({
			name = 'rust-analyzer',
			cmd = cmd,
			root_dir = root_dir,
			settings = settings,
			capabilities = capabilities,
		})
	end
})
```

### Expected behavior

_No response_

### Neovim version (nvim -v)

NVIM v0.10.0-dev-169+ge3f36377c

### Language server name/version

rust-analyzer 1.69.0 (84c898d 2023-04-16)

### Operating system/version

arch linux

### Log file

_No response_

### PR DESCRIPTION
**Pull request by nojnhuh**
This PR updates the file watcher used for LSP `workspace/didChangeWatchedFiles` to only poll files whose paths match at least one watcher from the LSP registration.

Fixes #23291

### CODE REVIEW

File: runtime/lua/vim/_watch.lua
Comment (glepnir) [inline runtime/lua/vim/_watch.lua:113]:
```suggestion
    local matches = vim.iter(opts.include_patterns):any(function(pattern)
        return path:match(pattern) )
```

File: runtime/lua/vim/lsp/_watchfiles.lua
Comment (glepnir) [inline runtime/lua/vim/lsp/_watchfiles.lua:271]:
```suggestion
        include_patterns[#include_patterns + 1] =  p
```

File: runtime/lua/vim/lsp/_watchfiles.lua
Comment (glepnir) [inline runtime/lua/vim/lsp/_watchfiles.lua:269]: maybe there can also use `vim.iter` syntax suger make it simply.

**Comment by nojnhuh** [comment]
@clason Since I'll update this to use `vim.iter` which doesn't exist in 0.9, should I open a separate PR to the release branch manually that doesn't use `vim.iter`? Or can I amend the automated backport PR?

File: runtime/lua/vim/lsp/_watchfiles.lua
Comment (nojnhuh) [reply runtime/lua/vim/lsp/_watchfiles.lua:271]: Out of curiosity, what's the difference between these two? There are a number of places that call `table.insert` like this so I'm wondering if those should be updated too.

**Comment by clason** [comment]
If you plan to use `vim.iter`, we can't backport; I'll remove the label. (I just thought "free performance, yay!" is a reason to backport.)

**Comment by clason** [comment]
But what you can do is keep the `vim.iter` changes in a separate commit so we can cherry-pick just the first one -- if you think this is worth it. I consider a significant editing performance regression like in the linked issue a bug that should be fixed in production as soon as possible.

**Comment by nojnhuh** [comment]
Sounds good. This PR currently should work with 0.9 so I'll make sure to keep any incompatible changes in a different commit.

**Comment by mfussenegger** [comment]
> I consider a significant editing performance regression like in the linked issue a bug that should be fixed in production as soon as possible.

Just as fyi: It's not enabled by default on 0.9 in part due to the performance issues.

**Comment by clason** [comment]
Ah, I missed that. Then we don't need to backport, at all.

**Comment by nojnhuh** [comment]
I think this change might still be valuable to backport for those users who choose to enable the feature in 0.9.

File: runtime/lua/vim/lsp/_watchfiles.lua
Comment (nojnhuh) [reply runtime/lua/vim/lsp/_watchfiles.lua:269]: @glepnir I took a look through `vim.iter` but couldn't come up with anything that seemed to make this much simpler. Did you have an idea of what this could possibly look like?

File: runtime/lua/vim/lsp/_watchfiles.lua
Comment (glepnir) [reply runtime/lua/vim/lsp/_watchfiles.lua:271]: inline expression fast in `table.insert` in luajit when you do it in a for loop

File: runtime/lua/vim/lsp/_watchfiles.lua
Comment (glepnir) [reply runtime/lua/vim/lsp/_watchfiles.lua:269]: if backport to 0.9 .ignore this :)

**Comment by nojnhuh** [comment]
Pushed one tiny change from @glepnir's feedback. I think I'll hold off on adding any `vim.iter` for now since it seems like making this easier to backport is worth it.

**Review by lewis6991 — APPROVED** [review]

**Comment by mfussenegger** [comment]
Just tried this and it again causes a significant editor freeze when lsp starts.

I wonder if we can off-load some of the file watching into a dedicated thread. I didn't follow how usable `uv.new_thread()` nowadays

**Comment by nojnhuh** [comment]
ACK, I'll take a look.

I also saw #23375 does something interesting where it looks like it does all its work on the main thread but chunks it up in a way that doesn't lock up everything else until it finishes. That approach seems like it might work here? That wouldn't make it complete any faster but would at least not freeze up the editor.

**Comment by mfussenegger** [comment]
> That approach seems like it might work here? 

yes that might also be worth exploring, but I'd tend to first do some profiling to see where the hotspots are and if there are some other options.

Another thing to look into is if we can avoid scheduling onto the main thread to begin with, given that the uv polling happens in the uv loop. E.g. is it safe to remove stuff like:

https://github.com/neovim/neovim/blob/4e5061dba765df2a74ac4a8182f6e7fe21da125d/runtime/lua/vim/_watch.lua#L108

**Comment by folke** [comment]
I'm running into this exact problem. Neovim is taking pretty much 100% cpu all the time in one of my projects.

It's a nodejs project that has 71000 files (node_modules etc).

What I don't get is why `fs_new_event` is **NOT** being used on linux?

https://github.com/neovim/neovim/blob/3ae62c470634e30c55d822c452af135fee50ddd5/runtime/lua/vim/lsp/_watchfiles.lua#L169

On Linux, all 71000 files are being polled instead. If I change the function to use watch, CPU usage drops to 1%.

In my case, the lsp actually asks to watch 1 file. The current implementation changes this to monitor recursively all the files in my project and then uses the one file as a pattern to match for changes.

**Comment by mfussenegger** [comment]
> What I don't get is why fs_new_event is NOT being used on linux?

See https://github.com/neovim/neovim/pull/22405#issuecomment-1445204284


> In my case, the lsp actually asks to watch 1 file. 


Can you share the `registerCapability` payload sent by the server?

**Comment by folke** [comment]
I still wouldn't use polling on linux. That literally kills performance.

I have two registrations:

```lua
{
        id = "rome_did_change_workspace_settings",
        method = "workspace/didChangeWatchedFiles",
        registerOptions = {
          watchers = { {
              globPattern = "/home/folke/projects/nova/rome.json",
              kind = 7
            } }
        }
      }
```

```lua
{
        id = "51859b96-121a-4eae-9c2a-dc44527dd4bb",
        method = "workspace/didChangeWatchedFiles",
        registerOptions = {
          watchers = { {
              globPattern = "**/{tailwind,tailwind.config,tailwind.*.config,tailwind.config.*}.{js,cjs,ts,mjs}"
            }, {
              globPattern = "**/{package-lock.json,yarn.lock,pnpm-lock.yaml}"
            }, {
              globPattern = "**/*.{css,scss,sass,less,pcss}"
            } }
        }
      }
```

The Rome one also scans all the files.

We should really never use polling.

**Comment by mfussenegger** [comment]
> We should really never use polling.

You are welcome to make a PR that uses watch + addresses the problems outlined in the comment

**Comment by folke** [comment]
That comment you linked to is about a potential wrong watched file event that could be generated when a subdirectory gets deleted.

It's not critical at all if in some cases a wrong event would be sent to the lsp server. This is due to a bug in libuv, so it will eventually be fixed there.

The work-around makes Neovim unusable for pretty much anyone that is working on a javascript project due to `node_modules`.

It would be much better to always use `fs_new_event` and just be aware that in the case of deleting a sub-directory, a wrong event might be sent to the lsp server.

Making Neovim use 100% cpu, to work around a bug in libuv doesn't feel like a proper fix for that issue.

**Comment by lewis6991** [comment]
Fwiw this is unusable for me on a python project so +1 for disabling by default

**Comment by folke** [comment]
@lewis6991 

```lua
require("vim.lsp._watchfiles")._watchfunc = require("vim._watch").watch
```

**Comment by mfussenegger** [comment]
> @lewis6991
> 
> ```lua
> require("vim.lsp._watchfiles")._watchfunc = require("vim._watch").watch
> ```
> 
> CPU **110%** => **1%**

This also breaks the functionality as it's not recursive on linux. 

To recap:

- We know there are performance issues. Part of the motivation to turn it on in _nightly_ was to get more feedback about them. Information that is particularly helpful:  Which Language server, `registerCapability` payload & rough outline of the folder structure/project size
- There is an easy workaround: Turn off the functionality
- The plan is to either improve the performance before 0.10 or turn it off again.

I'd appreciate it if we kept the discussion on a constructive level.

**Comment by folke** [comment]
> * There is an easy workaround: Turn off the functionality

Currently, this can't be disabled with some lsp servers that mis-behave.

Lsp servers like tailwindcss and rome register the capability even when dynamic registration is not enabled for watch files, so there should at least be an additional check in the `client/registerCapability` handler that dynamic registration is enabled.

**Comment by folke** [comment]
I wasn't aware that recursive  `fs_new_event` doesn't work on linux.

As an alternative approach, couldn't we just only trigger changes based on files changed within Neovim?

Or at least make that configurable? `watcher = "neovim" | "fs"` or something like that.
 
So with a `FileWritePost` autocmd, checking if the file matches any of the watchers and send an event only then.

## Pros:

- super fast
- easy to implement
- covers most use-cases

## Cons:

- when having multiple Neovim instances in the same directory, changes will only be sent from where the change was made
- changes made to those files outside of Neovim will not be triggered.

**Comment by mfussenegger** [comment]
> Currently, this can't be disabled with some lsp servers that mis-behave.
> Lsp servers like tailwindcss and rome register the capability even when dynamic registration is not enabled for watch files, so there should at least be an additional check in the client/registerCapability handler that dynamic registration is enabled.

I think that's a spec violation that should probably be fixed in those servers.
That said, I think I'd be okay with adding an additional check in the client.


> As an alternative approach, couldn't we just only trigger changes based on files changed within Neovim?


Changes that you make to files in Neovim are directly communicated to the server anyways. (Via `didChange`, based on `on_lines` of `nvim_buf_attach`). I think there would be little reason for a language server to rely on registerd file watchers for files the user is editing directly.

The main value proposition of `didChangeWatchedFiles` is that it works for changes made outside of Neovim. E.g. if you switch branches via git. Or to notice changes not related to a file, but e.g. if you create a directory.
If we take that away, I don't think there's anything left the functionality would provide.



@nojnhuh maybe we should bring back the default ignores (git, node_modules, etc) that you had in your initial PR. That would be a low hanging fruit.

**Comment by folke** [comment]
That explains a lot. Then at least for the lsp servers I'm using the watcher is indeed not really useful. tailwind watches src files and rome watches their config file. I only make changes to those files from within Neovim.

The default ignore would most likely fix my performance issues as well indeed.

**Comment by nojnhuh** [comment]
I just opened #23689 to address the erroneous registrations by servers when the capability isn't set.

I'll plan to add the default list of ignored file patterns to this PR since that's closely related. I still have some digging to do to figure out how to do this without the stuttering @mfussenegger was seeing at startup. I see the same behavior in my normal use case.

**Comment by clason** [comment]
@nojnhuh just to be clear: are you planning on making changes to _this_ PR, or can this be merged (and backported) already?

**Comment by nojnhuh** [comment]
Sorry, yes there are changes I'm planning to make for this PR yet. I've been making decent progress but it still needs some work.

**Comment by folke** [comment]
Another relevant issue: https://github.com/neovim/neovim/issues/23725#issuecomment-1561531718

`target` dircetory for rust projects

**Comment by nojnhuh** [comment]
Update on this: I recently learned about LPeg and figured that might be a good way to parse and match glob patterns. I don't expect matching an LPeg pattern to be much faster than matching a Lua pattern, but since LPeg natively supports alternatives (match "this" or "that"), then we can output a single LPeg pattern from a glob instead of a number of Lua patterns. My very unscientific testing shows this to be quite a bit faster at startup with the rest of the changes in this PR. Since those changes are mostly orthogonal to these though, I elected to open that here in a separate PR: #23788. I can definitely close that and make those changes here too if that would be easier to keep track of.

**Comment by nojnhuh** [comment]
I just pushed the changes to make this use LPeg to match both files to include and to exclude from polling. My rough testing shows this drops idle CPU usage way down for extreme cases like the one described in #23291 and even starts up a bit faster there.

The Go repo I tend to work on most doesn't have too many unwatched files and this change makes polling start a bit more slowly (~30ms -> ~100ms), but not enough to impact usability for me. If startup becomes unusably slow for larger repos where it wasn't already, we might need to start looking for ways to chunk up the work or offload to a separate coroutine/thread.

File: runtime/lua/vim/lsp/_watchfiles.lua
Comment (nojnhuh) [inline runtime/lua/vim/lsp/_watchfiles.lua:113]: @folke I wonder if exposing a variable like this would make it possible to exclude additional patterns per language, like `target` for rust.

Comment (mfussenegger) [reply runtime/lua/vim/lsp/_watchfiles.lua:113]: I think it's okay to expose this while keeping it private for us to be able to quickly test out stuff, but if we realize that this is useful to customize more often and before this starts popping up as recommendation to uers we should instead add a new setting for the client.

**Review by mfussenegger — APPROVED** [review]
The startup overhead that I had initially with this PR is gone, so from my side it's good to go

@folke @lewis6991 could you maybe also do a quick check to see if this improves things for you?

**Comment by lewis6991** [comment]
Unfortunately this doesn't fix my issues and I still get the error:

```
Error executing vim.schedule lua callback: ...sers/lewrus01/external/neovim/runtime/lua/vim/_watch.lua:132: ELOOP
stack traceback:
        [C]: in function 'assert'
        ...sers/lewrus01/external/neovim/runtime/lua/vim/_watch.lua:132: in function ''
        vim/_editor.lua: in function <vim/_editor.lua:0>
```

**Comment by nojnhuh** [comment]
@lewis6991 Is that an error you're not also seeing on master? If it's related to symlinks like the libuv docs suggest, I'd be a bit surprised if these changes were causing that. If you can put together an example repro that would be appreciated.

**Comment by lewis6991** [comment]
Yes it's on master too. This is with a python project with pyright. This is also running on a VM that doesn't have a huge amount of resources and has slow storage.

**Comment by lewis6991** [comment]
Just a data point, but these are the registration objects coming from pyright:

```
{
  id = "61acfbe9-e21b-4a1d-9676-6446bbe4d243",
  method = "workspace/didChangeWatchedFiles",
  registerOptions = {
    watchers = { {
        globPattern = "**/pyrightconfig.json",
        kind = 7
      }, {
        globPattern = "**",
        kind = 7
      } }
  }
}
{
  id = "c4e0f625-8a06-433f-93bc-b764e6df8c6f",
  method = "workspace/didChangeWatchedFiles",
  registerOptions = {
    watchers = { {
        globPattern = "**/pyrightconfig.json",
        kind = 7
      }, {
        globPattern = "**",
        kind = 7
      } }
  }
}
{
  id = "42720588-6689-4996-b9f7-4578fe8a0da1",
  method = "workspace/didChangeWatchedFiles",
  registerOptions = {
    watchers = { {
        globPattern = "**/pyrightconfig.json",
        kind = 7
      }, {
        globPattern = "**",
        kind = 7
      } }
  }
}

```

**Comment by nojnhuh** [comment]
I took a look and do see stuttering in this scenario.

> The above launches 10000 poll handles. For other virtualenvs I use this can be much more.
>
> The problem is that every file in .venv is being watched. Regardless of whether it is used or not.

I'm not sure there's an obvious fix for this. pyright asks for every file in every directory to be watched, so for large projects that will necessarily turn into a lot of polling handles being created. Even if I try moving the venv outside the project directory, pyright asks to watch it anyway:

```
["workspace/didChangeWatchedFiles"] = { {
          id = "e766283c-9069-41a5-b928-68a7a6246a28",
          method = "workspace/didChangeWatchedFiles",
          registerOptions = {
            watchers = { {
                globPattern = "**/pyrightconfig.json",
                kind = 7
              }, {
                globPattern = "**",
                kind = 7
              }, {
                globPattern = {
                  baseUri = "file:///home/jon/.venv/top/lib/python3.10/site-packages",
                  pattern = "**"
                },
                kind = 7
              }, {
                globPattern = {
                  baseUri = "file:///home/jon/.venv/top/lib64/python3.10/site-packages",
                  pattern = "**"
                },
                kind = 7
              } }
          }
        } }
```

If there's an obvious optimization here to more narrowly specify what should be watched, then it should be pyright handling that in the registration IMO.

> I think the following exclude just isn't sufficient for general use:

I envision the default as being a starting point that is meant to be generally applicable that plugins and users can augment with more language- or project-specific patterns to ignore. If this turns into a client setting like @mfussenegger mentioned, then it will probably be easier to integrate right into lspconfig.

**Comment by justinmk** [comment]
> pyright asks for every file in every directory to be watched, so for large projects that will necessarily turn into a lot of polling handles being created. Even if I try moving the venv outside the project directory, pyright asks to watch it anyway:

yikes, that is a disaster. 

> If there's an obvious optimization here to more narrowly specify what should be watched, then it should be pyright handling that in the registration IMO.

Agreed. But we at least need "visibility" for troubleshooting. Too-many file watchers can be a huge source of problems. 

1. At minimum (this PR) can we show a warning if the file watcher count exceeds, say, 1000?
2. Ideally we would also warn in `:checkhealth`
    - That implies something like a `vim.lsp.inspect()` that reports information about file watcher count (and other things such as info about all the random caches in the various vim.lsp modules)

**Comment by lewis6991** [comment]
Just an idea... but can we implement directory watchers using a single poll object?

If `foo/bar/baz/file` is modified, that should trigger the poll object on `foo`, `bar` and `baz` as `mtime` should get updated. Therefore if we need to watch `foo/**` we can place a single poll object on `foo` and when that triggers, we just need to scan `foo` (via `fs_stat`) to figure out which files actually changed.

The scanning can also be done asynchronously across several event loop iterations so it doesn't block the editor.

This is what I do in Gitsigns which watches the `.git` directory.

**Comment by mfussenegger** [comment]
@lewis6991 if I understood you right, this PR now doesn't regress the behavior for you compared to master, right?

If so, could we get this in and iterate further from there? I have the impression that this is already an improvement over what we got in master and your suggestions (which are good), can be built on top. They don't conceptually conflict with the changes here if I understood things right?

**Comment by lewis6991** [comment]
It is an improvement, but I'm also sensing we need to significantly rewrite large parts of this to get to an acceptable level of performance.

**Comment by mfussenegger** [comment]
> , but I'm also sensing we need to significantly rewrite large parts of this

Maybe, but I think at least the concept of ignoring well-known stuff like `.git/objects` could stay, even if the change detection gets optimized. So I'd go ahead and merge this if there are no objections. 

> 10000 watchers on a simple python project is just crazy.

Agreed

**Comment by lewis6991** [comment]
Ok makes sense.

**Comment by vlada-dudr** [comment]
> If `foo/bar/baz/file` is modified, that should cause a modification to `foo`, `bar` and `baz` as `mtime` should get updated. Therefore if we need to watch `foo/**` we can place a single poll object on `foo` and when that triggers, we just need to scan `foo` (via `fs_stat`) to figure out which files actually changed.

I am afraid that it is git specific behavior. And I couldn't find any documentation for it. Normally directory `mtime` on directory gets updated only when you change it, I.e. you add, remove or rename its child. `mtime` changes are not recursive normally.  That would be major performance issue.

I think the proper way to implement recursive file watching is through recursive inotify. Yeah, it is racy, but it shouldn't matter if neovim emits duplicate event. And if duplicate events are an issue, implementing debounce will solve it and improve performance for other platforms also.

In case neovim is not afraid to introduce new dependency, it is possible to use `inotify-tools` which ships c library which already solved this pain point. I don't know how easy it is to integrate it into uv loop.

**Comment by justinmk** [comment]
> it is possible to use `inotify-tools` which ships c library which already solved this pain point. I don't know how easy it is to integrate it into uv loop

Is that cross-platform? Is there any hope (pending issue/PR) of libuv addressing this itself?

**Comment by vlada-dudr** [comment]
> Is that cross-platform?
Inotify is itself linux sperecific api, so no. Windows and MacOS handle it well already.  I don't know how much support Neovim gives to BSD and other platforms. 

> Is there any hope (pending issue/PR) of libuv addressing this itself?
There are few rejected ones. Libuv seems to not want implement recursive watches themselves.

**Comment by lewis6991** [comment]
https://github.com/emcrisostomo/fswatch might be our best bet. We can either add it as an official dep or make it an optional runtime dependency and fallback to our current implementation when it is not available.

**Comment by mortezadadgar** [comment]
Please consider backporting this into upcoming 0.9.2 release.

**Pull request merged**
